File: lib/services/chat_service.dart
import 'package:cwc/models/chat_model.dart';
import 'package:cwc/services/notification_service.dart';
import 'package:cwc/services/supabase_service.dart';
import 'package:uuid/uuid.dart';

/// Preview text shown in chat list / notifications for image messages.
String chatMessagePreview(ChatMessageModel message) {
  if (message.messageType == 'image') return '📷 Photo';
  if (message.messageType == 'file') return '📎 File';
  return message.message;
}

/// Chat Service
/// Handles all chat-related database operations
class ChatService {
  final _supabase = SupabaseService.client;
  final _uuid = const Uuid();

  /// Get or create a chat room between two users
  Future<ChatRoomModel> getOrCreateChatRoom({
    required String user1Id,
    required String user2Id,
    String? collaborationId,
    String? workspaceId,
  }) async {
    try {
      // Try to find existing chat room
      final existingRooms = await _supabase
          .from('chat_rooms')
          .select()
          .or('and(user1_id.eq.$user1Id,user2_id.eq.$user2Id),and(user1_id.eq.$user2Id,user2_id.eq.$user1Id)')
          .maybeSingle();

      if (existingRooms != null) {
        return ChatRoomModel.fromChatRoomMap(existingRooms);
      }

      // Get user details for caching
      final user1Data = await _supabase
          .from('users')
          .select('name, profile_image_url')
          .eq('id', user1Id)
          .maybeSingle();

      final user2Data = await _supabase
          .from('users')
          .select('name, profile_image_url')
          .eq('id', user2Id)
          .maybeSingle();

      // Create new chat room
      final chatRoom = ChatRoomModel(
        id: _uuid.v4(),
        user1Id: user1Id,
        user2Id: user2Id,
        user1Name: user1Data?['name'],
        user2Name: user2Data?['name'],
        user1ProfileImage: user1Data?['profile_image_url'],
        user2ProfileImage: user2Data?['profile_image_url'],
        createdAt: DateTime.now(),
        collaborationId: collaborationId,
        workspaceId: workspaceId,
      );

      final chatRoomData = chatRoom.toChatRoomMap();
      chatRoomData.removeWhere((key, value) => value == null);

      await _supabase.from('chat_rooms').insert(chatRoomData);

      return chatRoom;
    } catch (e) {
      throw Exception('Failed to get or create chat room: ${e.toString()}');
    }
  }

  /// Get all chat rooms for a user (direct + group rooms they belong to)
  Future<List<ChatRoomModel>> getUserChatRooms(String userId) async {
    try {
      final rows = await _supabase
          .from('chat_rooms')
          .select()
          .or('user1_id.eq.$userId,user2_id.eq.$userId')
          .order('last_message_at', ascending: false);

      final rooms = rows.map((r) => ChatRoomModel.fromChatRoomMap(r)).toList();

      // Include group rooms where the user is a member.
      final memberRows = await _supabase
          .from('chat_room_members')
          .select('chat_room_id')
          .eq('user_id', userId);
      final memberRoomIds = memberRows
          .map((m) => m['chat_room_id'] as String?)
          .whereType<String>()
          .toSet()
          ..removeAll(rooms.map((r) => r.id));

      if (memberRoomIds.isNotEmpty) {
        final groupRows = await _supabase
            .from('chat_rooms')
            .select()
            .inFilter('id', memberRoomIds.toList());
        rooms.addAll(groupRows.map((r) => ChatRoomModel.fromChatRoomMap(r)));
      }

      rooms.sort((a, b) {
        final at = a.lastMessageAt?.millisecondsSinceEpoch ?? 0;
        final bt = b.lastMessageAt?.millisecondsSinceEpoch ?? 0;
        return bt.compareTo(at);
      });
      return rooms;
    } catch (e) {
      throw Exception('Failed to fetch chat rooms: ${e.toString()}');
    }
  }

  /// Create (or return existing) group chat room for a collaboration.
  Future<ChatRoomModel> createGroupChatRoom({
    required String collaborationId,
    required String ownerId,
    required String name,
    required List<Map<String, String?>> members, // {id, name, image}
  }) async {
    // Return existing room for this collaboration if present.
    final existing = await _supabase
        .from('chat_rooms')
        .select()
        .eq('collaboration_id', collaborationId)
        .eq('room_type', 'group')
        .maybeSingle();

    ChatRoomModel room;
    if (existing != null) {
      room = ChatRoomModel.fromChatRoomMap(existing);
    } else {
      room = ChatRoomModel(
        id: _uuid.v4(),
        user1Id: ownerId,
        roomType: 'group',
        name: name,
        collaborationId: collaborationId,
        createdAt: DateTime.now(),
      );
      final data = room.toChatRoomMap();
      data.removeWhere((key, value) => value == null);
      await _supabase.from('chat_rooms').insert(data);
    }

    // Ensure all members are registered.
    for (final m in members) {
      final uid = m['id'];
      if (uid == null) continue;
      await addGroupMember(
        chatRoomId: room.id,
        userId: uid,
        userName: m['name'],
        userProfileImage: m['image'],
      );
    }
    return room;
  }

  /// Add a member to a group chat room (idempotent).
  Future<void> addGroupMember({
    required String chatRoomId,
    required String userId,
    String? userName,
    String? userProfileImage,
  }) async {
    try {
      await _supabase.from('chat_room_members').upsert({
        'chat_room_id': chatRoomId,
        'user_id': userId,
        if (userName != null) 'user_name': userName,
        if (userProfileImage != null) 'user_profile_image': userProfileImage,
      }, onConflict: 'chat_room_id,user_id');
    } catch (_) {
      // ignore duplicates
    }
  }

  /// Get the group chat room for a collaboration, if any.
  Future<ChatRoomModel?> getGroupRoomForCollaboration(String collaborationId) async {
    final result = await _supabase
        .from('chat_rooms')
        .select()
        .eq('collaboration_id', collaborationId)
        .eq('room_type', 'group')
        .maybeSingle();
    if (result == null) return null;
    return ChatRoomModel.fromChatRoomMap(result);
  }

  /// Group room members (with cached names).
  Future<List<Map<String, dynamic>>> getGroupMembers(String chatRoomId) async {
    final rows = await _supabase
        .from('chat_room_members')
        .select()
        .eq('chat_room_id', chatRoomId);
    return List<Map<String, dynamic>>.from(rows);
  }

  /// Get chat room by ID
  Future<ChatRoomModel?> getChatRoomById(String chatRoomId) async {
    try {
      final result = await _supabase
          .from('chat_rooms')
          .select()
          .eq('id', chatRoomId)
          .maybeSingle();

      if (result == null) return null;
      return ChatRoomModel.fromChatRoomMap(result);
    } catch (e) {
      throw Exception('Failed to fetch chat room: ${e.toString()}');
    }
  }

  /// Total unread messages across all chat rooms (direct + group).
  Future<int> getTotalUnreadCount(String userId) async {
    try {
      final rooms = await getUserChatRooms(userId);
      return rooms.fold<int>(
        0,
        (sum, room) => sum + room.getUnreadCount(userId),
      );
    } catch (_) {
      return 0;
    }
  }

  /// Send a message
  Future<String> sendMessage(ChatMessageModel message) async {
    try {
      final messageData = message.toMessageMap();
      messageData.removeWhere((key, value) => value == null);

      await _supabase
          .from('messages')
          .insert(messageData);

      // Update chat room with last message
      final chatRoom = await getChatRoomById(message.chatRoomId);
      if (chatRoom != null && chatRoom.isGroup) {
        await _supabase.from('chat_rooms').update({
          'last_message': '${message.senderName.split(' ').first}: ${chatMessagePreview(message)}',
          'last_message_at': message.createdAt.toIso8601String(),
          'updated_at': DateTime.now().toIso8601String(),
        }).eq('id', message.chatRoomId);

        // Notify all other group members.
        try {
          final members = await getGroupMembers(message.chatRoomId);
          final notificationService = NotificationService();
          for (final m in members) {
            final uid = m['user_id'] as String?;
            if (uid == null || uid == message.senderId) continue;
            await notificationService.sendChatMessageNotification(
              receiverUserId: uid,
              senderName: '${message.senderName} (${chatRoom.name ?? 'Team'})',
              message: chatMessagePreview(message),
              chatRoomId: message.chatRoomId,
            );
          }
        } catch (_) {}
      } else if (chatRoom != null) {
        final isUser1 = message.senderId == chatRoom.user1Id;
        final receiverId = isUser1 ? chatRoom.user2Id : chatRoom.user1Id;
        final updateData = {
          'last_message': chatMessagePreview(message),
          'last_message_at': message.createdAt.toIso8601String(),
          'updated_at': DateTime.now().toIso8601String(),
          if (isUser1) 'unread_count2': chatRoom.unreadCount2 + 1,
          if (!isUser1) 'unread_count1': chatRoom.unreadCount1 + 1,
        };

        await _supabase
            .from('chat_rooms')
            .update(updateData)
            .eq('id', message.chatRoomId);

        // Send notification to the receiver
        final notificationService = NotificationService();
        await notificationService.sendChatMessageNotification(
          receiverUserId: receiverId,
          senderName: message.senderName,
          message: chatMessagePreview(message),
          chatRoomId: message.chatRoomId,
        );
      }

      return message.id;
    } catch (e) {
      throw Exception('Failed to send message: ${e.toString()}');
    }
  }

  /// Get messages for a chat room
  Future<List<ChatMessageModel>> getChatMessages(String chatRoomId, {int limit = 50}) async {
    try {
      final rows = await _supabase
          .from('messages')
          .select()
          .eq('chat_room_id', chatRoomId)
          .order('created_at', ascending: false)
          .limit(limit);

      final messages = rows
          .map((m) => ChatMessageModel.fromMessageMap(m))
          .toList();

      // Reverse to show oldest first
      return messages.reversed.toList();
    } catch (e) {
      throw Exception('Failed to fetch messages: ${e.toString()}');
    }
  }

  /// Mark messages as read
  Future<void> markMessagesAsRead(String chatRoomId, String userId) async {
    try {
      // Mark all unread messages as read
      await _supabase
          .from('messages')
          .update({'is_read': true, 'updated_at': DateTime.now().toIso8601String()})
          .eq('chat_room_id', chatRoomId)
          .neq('sender_id', userId)
          .eq('is_read', false);

      // Reset unread count in chat room
      final chatRoom = await getChatRoomById(chatRoomId);
      if (chatRoom != null) {
        final isUser1 = userId == chatRoom.user1Id;
        final updateData = {
          if (isUser1) 'unread_count1': 0,
          if (!isUser1) 'unread_count2': 0,
          'updated_at': DateTime.now().toIso8601String(),
        };

        await _supabase
            .from('chat_rooms')
            .update(updateData)
            .eq('id', chatRoomId);
      }
    } catch (e) {
      throw Exception('Failed to mark messages as read: ${e.toString()}');
    }
  }

  /// Get stream of messages for real-time chat
  Stream<List<ChatMessageModel>> getMessagesStream(String chatRoomId) {
    return _supabase
        .from('messages')
        .stream(primaryKey: ['id'])
        .eq('chat_room_id', chatRoomId)
        .order('created_at', ascending: false)
        .limit(50)
        .map((data) {
          final messages = data
              .map((m) => ChatMessageModel.fromMessageMap(m))
              .toList();
          return messages.reversed.toList();
        });
  }

  /// Get stream of chat rooms for real-time updates
  Stream<List<ChatRoomModel>> getChatRoomsStream(String userId) {
    return _supabase
        .from('chat_rooms')
        .stream(primaryKey: ['id'])
        .map((data) {
          // Filter chat rooms where user is either user1 or user2
          final filteredData = data.where((room) {
            return room['user1_id'] == userId || room['user2_id'] == userId;
          }).toList();
          
          // Sort by last_message_at
          filteredData.sort((a, b) {
            final aTime = a['last_message_at'] != null 
                ? DateTime.parse(a['last_message_at']).millisecondsSinceEpoch 
                : 0;
            final bTime = b['last_message_at'] != null 
                ? DateTime.parse(b['last_message_at']).millisecondsSinceEpoch 
                : 0;
            return bTime.compareTo(aTime); // Descending order
          });
          
          return filteredData.map((r) => ChatRoomModel.fromChatRoomMap(r)).toList();
        });
  }

  /// Delete a message
  Future<void> deleteMessage(String messageId) async {
    try {
      await _supabase.from('messages').delete().eq('id', messageId);
    } catch (e) {
      throw Exception('Failed to delete message: ${e.toString()}');
    }
  }

  /// Delete a chat room (and all its messages)
  Future<void> deleteChatRoom(String chatRoomId) async {
    try {
      await _supabase.from('messages').delete().eq('chat_room_id', chatRoomId);
      await _supabase.from('chat_rooms').delete().eq('id', chatRoomId);
    } catch (e) {
      throw Exception('Failed to delete chat: ${e.toString()}');
    }
  }
}

File: cwc-web/src/services/chatService.js
import { supabase } from '../lib/supabase'
import { uuid } from '../lib/helpers'
import { notificationService } from './notificationService'

const preview = (type, msg) =>
  type === 'image' ? '📷 Photo' : type === 'file' ? '📎 File' : msg

export const chatService = {
  async getUserChatRooms(userId) {
    // Direct rooms where the user is a participant.
    const { data: direct, error } = await supabase
      .from('chat_rooms')
      .select('*')
      .or(`user1_id.eq.${userId},user2_id.eq.${userId}`)
      .order('last_message_at', { ascending: false, nullsFirst: false })
    if (error) throw error

    // Group rooms where the user is a member.
    const { data: memberships } = await supabase
      .from('chat_room_members')
      .select('chat_room_id')
      .eq('user_id', userId)
    const groupIds = (memberships || []).map((m) => m.chat_room_id)
    let groups = []
    if (groupIds.length) {
      const { data: g } = await supabase
        .from('chat_rooms')
        .select('*')
        .in('id', groupIds)
      groups = g || []
    }
    const all = [...(direct || []), ...groups]
    const map = {}
    for (const r of all) map[r.id] = r
    return Object.values(map).sort(
      (a, b) =>
        new Date(b.last_message_at || b.created_at) -
        new Date(a.last_message_at || a.created_at)
    )
  },

  async getOrCreateDirectRoom({ user1Id, user2Id, workspaceId, collaborationId }) {
    const { data: existing } = await supabase
      .from('chat_rooms')
      .select('*')
      .or(
        `and(user1_id.eq.${user1Id},user2_id.eq.${user2Id}),and(user1_id.eq.${user2Id},user2_id.eq.${user1Id})`
      )
      .limit(1)
      .maybeSingle()
    if (existing) return existing

    const { data: users } = await supabase
      .from('users')
      .select('id, name, profile_image_url')
      .in('id', [user1Id, user2Id])
    const u1 = users?.find((u) => u.id === user1Id)
    const u2 = users?.find((u) => u.id === user2Id)

    const { data, error } = await supabase
      .from('chat_rooms')
      .insert({
        id: uuid(),
        user1_id: user1Id,
        user2_id: user2Id,
        user1_name: u1?.name,
        user2_name: u2?.name,
        user1_profile_image: u1?.profile_image_url,
        user2_profile_image: u2?.profile_image_url,
        room_type: 'direct',
        workspace_id: workspaceId || null,
        collaboration_id: collaborationId || null,
        created_at: new Date().toISOString(),
      })
      .select()
      .single()
    if (error) throw error
    return data
  },

  async getMessages(roomId, limit = 100) {
    const { data, error } = await supabase
      .from('messages')
      .select('*')
      .eq('chat_room_id', roomId)
      .order('created_at', { ascending: false })
      .limit(limit)
    if (error) throw error
    return (data || []).reverse()
  },

  async getRoom(roomId) {
    const { data, error } = await supabase
      .from('chat_rooms')
      .select('*')
      .eq('id', roomId)
      .maybeSingle()
    if (error) throw error
    return data
  },

  async sendMessage({ room, senderId, senderName, senderImage, message, type = 'text', imageUrl }) {
    const text = type === 'image' ? '📷 Photo' : message
    const { data, error } = await supabase
      .from('messages')
      .insert({
        id: uuid(),
        chat_room_id: room.id,
        sender_id: senderId,
        sender_name: senderName,
        sender_profile_image: senderImage || null,
        message: text,
        message_type: type,
        image_url: imageUrl || null,
        is_read: false,
        created_at: new Date().toISOString(),
      })
      .select()
      .single()
    if (error) throw error

    const isGroup = room.room_type === 'group'
    const roomUpdate = {
      last_message: isGroup
        ? `${(senderName || '').split(' ')[0]}: ${preview(type, message)}`
        : preview(type, message),
      last_message_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
    }
    if (!isGroup) {
      if (room.user1_id === senderId) roomUpdate.unread_count2 = (room.unread_count2 || 0) + 1
      else roomUpdate.unread_count1 = (room.unread_count1 || 0) + 1
    }
    await supabase.from('chat_rooms').update(roomUpdate).eq('id', room.id)

    // notify receiver(s)
    try {
      if (isGroup) {
        const { data: members } = await supabase
          .from('chat_room_members')
          .select('user_id')
          .eq('chat_room_id', room.id)
        for (const m of members || []) {
          if (m.user_id !== senderId) {
            await notificationService.create({
              userId: m.user_id,
              title: room.name || 'New message',
              message: `${(senderName || '').split(' ')[0]}: ${preview(type, message)}`,
              type: 'chat_message',
              metadata: { chat_room_id: room.id, sender_name: senderName },
            })
          }
        }
      } else {
        const receiver = room.user1_id === senderId ? room.user2_id : room.user1_id
        await notificationService.create({
          userId: receiver,
          title: senderName || 'New message',
          message: preview(type, message),
          type: 'chat_message',
          metadata: { chat_room_id: room.id, sender_name: senderName },
        })
      }
    } catch (e) {
      /* ignore */
    }
    return data
  },

  async markRead(roomId, userId) {
    await supabase
      .from('messages')
      .update({ is_read: true })
      .eq('chat_room_id', roomId)
      .neq('sender_id', userId)
      .eq('is_read', false)
    const room = await this.getRoom(roomId)
    if (room && room.room_type !== 'group') {
      const patch =
        room.user1_id === userId ? { unread_count1: 0 } : { unread_count2: 0 }
      await supabase.from('chat_rooms').update(patch).eq('id', roomId)
    }
  },

  subscribeMessages(roomId, onInsert) {
    const channel = supabase
      .channel(`msgs-${roomId}`)
      .on(
        'postgres_changes',
        { event: 'INSERT', schema: 'public', table: 'messages', filter: `chat_room_id=eq.${roomId}` },
        onInsert
      )
      .subscribe()
    return () => supabase.removeChannel(channel)
  },

  subscribeRooms(userId, onChange) {
    const channel = supabase
      .channel(`rooms-${userId}`)
      .on('postgres_changes', { event: '*', schema: 'public', table: 'chat_rooms' }, onChange)
      .subscribe()
    return () => supabase.removeChannel(channel)
  },

  // group room for a collaboration
  async getGroupRoom(collaborationId) {
    const { data } = await supabase
      .from('chat_rooms')
      .select('*')
      .eq('collaboration_id', collaborationId)
      .eq('room_type', 'group')
      .maybeSingle()
    return data
  },
}
